build.rs

```diff
@@ -348,13 +348,28 @@ fn pregenerate_asm_main() {
         perlasm(&perlasm_src_dsts, asm_target);
 
         if asm_target.preassemble {
+            // Preassembly is currently only done for Windows targets.
+            assert_eq!(&asm_target.oss, &[WINDOWS]);
+            let os = WINDOWS;
+            let env = "msvc";
+
             if !std::mem::replace(&mut generated_prefix_headers, true) {
                 generate_prefix_symbols_nasm(&pregenerated, &ring_core_prefix()).unwrap();
             }
             let srcs = asm_srcs(perlasm_src_dsts);
+
+            let target = Target {
+                arch: asm_target.arch.to_owned(),
+                os: os.to_owned(),
+                env: env.to_string(),
+                is_git: true,
+                is_debug: false,
+            };
+
+            let compiler = cc::Build::new().get_compiler();
```

Comment:
Windows AArch64 (for now) will always be compiled with clang. Would we need to hardcode it here as well or, because preassembly is only expected to run by package.sh, we will need to set environment variables there?

I personally would prefer to keep that hardcoded compiler in a single place somewhere, so maybe we could introduce a new function to create compiler in one place? (Maybe this is something for our PR actually)

Reply:
> Windows AArch64 (for now) will always be compiled with clang. Would we need to hardcode it here as well or, because preassembly is only expected to run by package.sh, we will need to set environment variables there?
>
> I personally would prefer to keep that hardcoded compiler in a single place somewhere, so maybe we could introduce a new function to create compiler in one place? (Maybe this is something for our PR actually)

I think here we would add a check that `compiler.is_like_clang()`, and then change `mk/package.sh` to set `TARGET_CC` to the path to VS2019's clang.

Reply:
My concern in this case is that we'll have to put a check for Windows AArch64 in 2 places – package.sh as you've mentioned and build.rs, because the users of *ring* will still have to compile C code with Clang.

Reply:
On the other hand, it might be temporary only for C code, while for assembly it'll probably stay permanent 🤔

Reply:
> My concern in this case is that we'll have to put a check for Windows AArch64 in 2 places – package.sh as you've mentioned and build.rs, because the users of ring will still have to compile C code with Clang. Since it's a (hopefully) temporary thing, I'd suggest to have this knowledge in 1 place.

It would be nice to have this in one place but OTOH the process for `package` is very different than the normal build. For `package` ideally the inputs will become completely fixed: a specific version of clang from a specific download, similar to how `nasm` is done.

For the build, at least when building from crates.io, ideally we don't force any specific compiler on the user. It seems temporarily the build will only work if the compiler is clang or clang-cl, but whether it's in `$PATH` or not, or what that compiler is named (`clang` vs `clang-12`, for example) wouldn't be hard-coded in.

Reply:
Let's move the remaining discussion of aarch64-specific stuff into PR #1339.

Reply:
> but whether it's in `$PATH` or not, or what that compiler is named (`clang` vs `clang-12`, for example) wouldn't be hard-coded in.

I'm afraid that would make usage of *ring* unnecessarily over-complicated on Windows AArch64. I'd probably still suggest to hardcode `clang-cl` for now, for better compatibility and for simpler usage.

At the same time, if you don't want to release a new *ring* version before it's compilable under MSVC on AArch64, and given that you still need additional setup (Clang in the PATH), maybe it doesn't really matter much. 🤔

I guess I don't have a strong opinion but slightly inclined to hardcode `clang-cl` with a `FIXME` comment.

Reply:
Sorry, again didn't notice the last comment 🤦‍♂️  Will move the discussion to #1339

Reply:
> I'm afraid that would make usage of ring unnecessarily over-complicated on Windows AArch64. I'd probably still suggest to hardcode clang-cl for now, for better compatibility and for simpler usage.

Maybe we can do something like "if the target arch is aarch64 and the env is msvc and no compiler has been set then default to whataever `clang` there is on the path." In the products I work on we will set `TARGET_AARCH64_PC_WINDOWS_MSVC_CC` or similar to a specific version of clang.

> At the same time, if you don't want to release a new ring version before it's compilable under MSVC on AArch64, and given that one still needs additional setup (Clang in the $PATH), maybe it doesn't really matter much. 🤔

I suggest the following course of action, which seems ideal to me:
PR 1. Merge this.
PR 2. Make a PR to get the build working when `$TARGET_AARCH64_PC_WINDOWS_MSVC_CC` points to VS2019's `clang`, either a full path or "clang" when it is in `$PATH`.
PR 3. Change the code to compile with VS2019's cl.exe (or clang or clang-cl).

Note that defaulting to "clang" in "$PATH" creates some problems in some environments where people have a different `clang` in `$PATH`. In some projects I work on, this can happen because they need `clang` to target wasm32.

Reply:
Responded in #1339 😅